### ISSUE TITLE
Conflict between Variable Name and Type in Unresolvable Scope

### ISSUE THREAD

**Issue by luckyabsoluter** [godotengine/godot#92008]
### Tested versions

v4.2.1.stable.official [b09f793f5]
v4.2.2.stable.official [15073afe3]

### System information

N/A

### Issue description

When declaring a variable with the syntax `var a: TypeName` followed by `var TypeName` in Godot Engine, an error is raised stating "Local variable 'TypeName' cannot be used as a type." This issue arises due to an unresolvable scope conflict where the engine cannot distinguish between `TypeName` as both a type and a variable.

Example Code:
```gdscript
func my_func():
	var a: TypeName # error
	var TypeName
```

Error Message:
```
Local variable "TypeName" cannot be used as a type.
```

The engine should either:
1. Allow the declaration without raising an error, correctly differentiating between the type and the variable name.
2. Provide a more informative error message that clearly explains the conflict and suggests using a different variable name.

Also,
```gdscript
var Variant
func my_func():
	var a: Variant # no error
```
```gdscript
func my_func():
	var Variant
	var a: Variant # error
```
```gdscript
func my_func():
	var a: Variant # error
	var Variant
```

### Steps to reproduce

```gdscript
func my_func():
	var a: Variant # error
	var Variant
```

### Minimal reproduction project (MRP)

N/A

**Comment by RedMser** [comment]
> Allow the declaration without raising an error, correctly differentiating between the type and the variable name.

I don't think this is possible, since you can still call static functions that exist on the types (not sure if Godot currently has any?). Also for classes, it wouldn't work due to static functions and methods like .new()

**Comment by luckyabsoluter** [comment]
@RedMser
It seems there is a misunderstanding. My argument is not that variables should be used as types. Rather, my point is that there should be no conflict between variable names and type names within the same scope. Therefore, variables declared below should not be referenced above.

To clarify, please see the following example:
```gdscript
const Apple = preload("res://apple.gd")
func my_func() :
	var my_var: Apple
	var Apple = "pineapple"
	print(Apple)
```
The `Apple` in `var my_var: Apple` does not reference a variable. Similarly, the `Variant` in `var my_var: Variant` is also not a variable.
It's a discussion about shadowing, an unpreferred coding practice where an identifier is covered.
Please let me know if I am misunderstood.

**Comment by dalexeev** [comment]
> This issue arises due to an unresolvable scope conflict where the engine cannot distinguish between `TypeName` as both a type and a variable.

A local constant can be used as a type hint if it contains a preloaded class or enum. So while we don't support types as first-class citizens, types share the same scope as variables, constants, functions, etc. So if you declare a local identifier, then you can no longer use the same name class or global identifier, no matter what it is: a variable, a type, etc.

The only bug here is that a constant declared after use should not obscure the identifier, only produce a warning. We fixed a similar bug in #79880, I'll look at that later for types.

Also note that we recommend using `PascalCase` only for types and constants that contain a preloaded type. Local variables should use `snake_case`, so the example would not be possible if you followed the recommended style. Shadowing is also generally seen as a bad practice.
